### ISSUE TITLE
java.lang.OutOfMemoryError: Failed to allocate a 8431340 byte allocation with 210496 free bytes and 205KB until OOM

### ISSUE THREAD

**Issue by jaredsburrows** [PhilJay/MPAndroidChart#1386]
Version: `compile 'com.github.PhilJay:MPAndroidChart:v2.1.6'`

I was running monkey on an app for 2+ hours and ran into a memory issue with the charting library:

```
// CRASH: <package> (pid 29580)
// Short Msg: java.lang.OutOfMemoryError
// Long Msg: java.lang.OutOfMemoryError: Failed to allocate a 8431340 byte allocation with 210496 free bytes and 205KB until OOM
// Build Label: google/shamu/shamu:6.0.1/MMB29S/2489379:user/release-keys
// Build Changelist: 2489379
// Build Time: 1450217770000
// java.lang.OutOfMemoryError: Failed to allocate a 8431340 byte allocation with 210496 free bytes and 205KB until OOM
// 	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
// 	at android.graphics.Bitmap.nativeCreate(Native Method)
// 	at android.graphics.Bitmap.createBitmap(Bitmap.java:831)
// 	at android.graphics.Bitmap.createBitmap(Bitmap.java:808)
// 	at android.graphics.Bitmap.createBitmap(Bitmap.java:775)
// 	at com.github.mikephil.charting.renderer.LineChartRenderer.drawData(LineChartRenderer.java:88)
// 	at com.github.mikephil.charting.charts.BarLineChartBase.onDraw(BarLineChartBase.java:247)
// 	at android.view.View.draw(View.java:16184)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15180)
// 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
// 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15140)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.support.design.widget.CoordinatorLayout.drawChild(CoordinatorLayout.java:1077)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.support.v4.widget.DrawerLayout.drawChild(DrawerLayout.java:1229)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15175)
// 	at android.view.View.draw(View.java:15954)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
// 	at android.view.View.draw(View.java:16187)
// 	at com.android.internal.policy.PhoneWindow$DecorView.draw(PhoneWindow.java:2690)
// 	at android.view.View.updateDisplayListIfDirty(View.java:15180)
// 	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:281)
// 	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:287)
// 	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:322)
// 	at android.view.ViewRootImpl.draw(ViewRootImpl.java:2615)
// 	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2434)
// 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2067)
// 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
// 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
// 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
// 	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
// 	at android.view.Choreographer.doFrame(Choreographer.java:606)
// 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
// 	at android.os.Handler.handleCallback(Handler.java:739)
// 	at android.os.Handler.dispatchMessage(Handler.java:95)
// 	at android.os.Looper.loop(Looper.java:148)
// 	at android.app.ActivityThread.main(ActivityThread.java:5417)
// 	at java.lang.reflect.Method.invoke(Native Method)
// 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
// 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
// 
** Monkey aborted due to error.
Events injected: 256961
:Sending rotation degree=0, persist=false
:Dropped: keys=0 pointers=814 trackballs=0 flips=0 rotations=0
## Network stats: elapsed time=8392464ms (0ms mobile, 0ms wifi, 8392464ms not connected)
** System appears to have crashed at event 256961 of 999999 using seed 5566
```

**Comment by jaredsburrows** [comment]
Other OOM errors:
- https://github.com/PhilJay/MPAndroidChart/issues/1348
- https://github.com/PhilJay/MPAndroidChart/issues/932
- https://github.com/PhilJay/MPAndroidChart/issues/1197

**Comment by PhilJay** [comment]
A lot of people are running into OOM, and that does not surprise me because especially `LineChart` allocates a lot of memory when drawing with bitmaps.

However, and OOM does not mean that the library has a memory leak. It simply means that it tried to allocate memory that was not available. 
This could just as well mean that some other part in your application is leaking memory, and the library (beacause it allocates large chunks of memory) is the first thing to crash because of that.

I have focused on the issue for quite some time now, and I believe that the chart does everything it should not to leak memory:

```java
public void releaseBitmap() {
        if (mDrawBitmap != null) {
            mDrawBitmap.recycle();
            mDrawBitmap = null;
        }
    }
```

**Comment by jaredsburrows** [comment]
@PhilJay Thanks for your reply! I was trying aggregate all of the issues, cross referencing them to try and find some solution. I will re-run the monkey tests with leak canary and try to trace these memory issues further.

I am currently displaying the `LineChart` on a `Fragment` but I am not `retaining` it or `adding` it to the stack. I am using `replace`. I do not want to use `System.gc()` or increase the app's heap.

**Comment by brcinho** [comment]
The releaseBitmap() method only gets called in onDetachedFromWindow() and I don't see where and when is that method invoked. On the other hand, onDetachedFromWindow() is called only in LineChart and PieChart, what about other types of chart?

**Comment by brcinho** [comment]
In issue #1197 i think I explained the potential root of the problem, but I don't know how it got ignored for such a long time.

**Comment by PhilJay** [comment]
You got ignored for such a long time because I have a life and a job besides this and about 5 new issues every day plus dozens of emails. I'm looking into this.

**Comment by brcinho** [comment]
You are absolutely right. I went out of line there and therefore I apologize. If you need help and want me to jump in with trying to fix it and submitting a pull request, let me know.

**Comment by PhilJay** [comment]
I just replaced the standart references to `Bitmaps` by `WeakReference`. Could you please try if this makes any difference?

**Comment by jaredsburrows** [comment]
@brcinho I am sure @PhilJay would be more than happy to accept pull request referencing this issue.

@PhilJay Yes, I can test this build. Via jitpack? I am using a different computer at the moment and I can't submit a PR but I believe you should check for null on the `WeakReference` as you are not guaranteed it will be there:

From:
```
if (mDrawBitmap == null
 -                || (mDrawBitmap.getWidth() != width)
 -                || (mDrawBitmap.getHeight() != height)) {
```
To:
```
if (mDrawBitmap == null
 -                || (mDrawBitmap.get() != null && mDrawBitmap.getWidth() != width)
 -                || (mDrawBitmap.get() != null && mDrawBitmap.getHeight() != height)) {
```

**Comment by PhilJay** [comment]
Yes you can test via jitpack just use the latest commit and not release

**Comment by jaredsburrows** [comment]
@PhilJay I am going to try, https://jitpack.io/#PhilJay/MPAndroidChart/aa5730d19f:
```
dependencies {
	        compile 'com.github.PhilJay:MPAndroidChart:aa5730d19f'
}
```

Hopefully `monkey` tests can run past 2 hours. I will return with my results.

**Comment by PhilJay** [comment]
Awesome, any results yet?

**Comment by jaredsburrows** [comment]
@PhilJay I am able to get LeakCanary to crash the app from too much memory using the `monkey` binary running an event every single `0.100` a second: `adb -d shell monkey -p com.yahoo.flurry  --monitor-native-crashes --pct-trackball 0 --pct-nav 0 --pct-appswitch 0 --pct-syskeys 0 --pct-anyevent 0 -s 5566 --throttle 100 -v -v -v 999999 -f` and `0.500` a second: `adb -d shell monkey -p com.yahoo.flurry  --monitor-native-crashes --pct-trackball 0 --pct-nav 0 --pct-appswitch 0 --pct-syskeys 0 --pct-anyevent 0 -s 5566 --throttle 500 -v -v -v 999999 -f`

Without LeakCanary, I can get `monkey` to run for 2hrs+. The reason I mention leak canary is because if it is able to detect memory increases then maybe there is still a memory issue. I honestly believe that if I can get monkey to run 5hrs+ on an app with the `LineChart`, then we should be fine. I will continue to report back.

**Comment by jaredsburrows** [comment]
@PhilJay I just ran `adb monkey` and was able to reproduce the memory leak:

```
// CRASH: package (pid 6447)
// Short Msg: java.lang.OutOfMemoryError
// Long Msg: java.lang.OutOfMemoryError: Failed to allocate a 4842156 byte allocation with 2271684 free bytes and 2MB until OOM
// Build Label: htc/verizon_wwe/dlx:4.4.2/KOT49H/322791.5:user/release-keys
// Build Changelist: 95cf91ea4d
// Build Time: 1426394589000
// java.lang.OutOfMemoryError: Failed to allocate a 4842156 byte allocation with 2271684 free bytes and 2MB until OOM
// 	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
// 	at android.graphics.Bitmap.nativeCreate(Native Method)
// 	at android.graphics.Bitmap.createBitmap(Bitmap.java:817)
// 	at android.graphics.Bitmap.createBitmap(Bitmap.java:794)
// 	at android.graphics.Bitmap.createBitmap(Bitmap.java:761)
// 	at package(SourceFile:89)
// 	at com.github.mikephil.charting.charts.BarLineChartBase.onDraw(SourceFile:247)
// 	at android.view.View.draw(View.java:15114)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14048)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3388)
// 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3367)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14008)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.support.design.widget.CoordinatorLayout.drawChild(SourceFile:1077)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.support.v4.widget.DrawerLayout.drawChild(SourceFile:1229)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14043)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.View.draw(View.java:14838)
// 	at android.view.ViewGroup.drawChild(ViewGroup.java:3404)
// 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3198)
// 	at android.view.View.draw(View.java:15117)
// 	at android.widget.FrameLayout.draw(FrameLayout.java:592)
// 	at com.android.internal.policy.impl.PhoneWindow$DecorView.draw(PhoneWindow.java:2850)
// 	at android.view.View.updateDisplayListIfDirty(View.java:14048)
// 	at android.view.View.getDisplayList(View.java:14071)
// 	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:266)
// 	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:272)
// 	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:311)
// 	at android.view.ViewRootImpl.draw(ViewRootImpl.java:2499)
// 	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2337)
// 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1968)
// 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1054)
// 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:5786)
// 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:767)
// 	at android.view.Choreographer.doCallbacks(Choreographer.java:580)
// 	at android.view.Choreographer.doFrame(Choreographer.java:550)
// 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:753)
// 	at android.os.Handler.handleCallback(Handler.java:739)
// 	at android.os.Handler.dispatchMessage(Handler.java:95)
// 	at android.os.Looper.loop(Looper.java:135)
// 	at android.app.ActivityThread.main(ActivityThread.java:5256)
// 	at java.lang.reflect.Method.invoke(Native Method)
// 	at java.lang.reflect.Method.invoke(Method.java:372)
// 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:898)
// 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:693)
// 
** Monkey aborted due to error.
Events injected: 26967
:Sending rotation degree=0, persist=false
:Dropped: keys=0 pointers=199 trackballs=0 flips=0 rotations=0
## Network stats: elapsed time=1510034ms (0ms mobile, 0ms wifi, 1510034ms not connected)
** System appears to have crashed at event 26967 of 999999 using seed 5566
```

In my original post:
```
//  at com.github.mikephil.charting.renderer.LineChartRenderer.drawData(LineChartRenderer.java:88)
//  at com.github.mikephil.charting.charts.BarLineChartBase.onDraw(BarLineChartBase.java:247)
```

In this post:
```
// 	at com.github.mikephil.charting.charts.BarLineChartBase.onDraw(SourceFile:247)
```

**Comment by jaredsburrows** [comment]
Just now, only 30mins+ the `monkey`, I was able to get my application to ANR:

```
----- end 232 -----
// anr traces status was 0
meminfo:
Applications Memory Usage (kB):
Uptime: 6942369 Realtime: 6942357

Total PSS by process:
   208243 kB: package (pid 17716 / activities)   <----
    89164 kB: system (pid 540)
    57887 kB: com.android.systemui (pid 759)
```

**Comment by jaredsburrows** [comment]
@PhilJay I am trying to give as much information as possible. Is there anything I can try or do on my side? Should I try "removing all views" in `onDestroy`? Should I invalidate the graph in `onDestroy`? Should I remove all data sets entry points in `onDestroy`?

**Comment by dalewking** [comment]
So regarding the failure to allocate this bitmap, I happened to notice (and was coming here to report) that the renderers are creating this bitmap but IT IS NEVER ACTUALLY USED!!!! All the drawing operations are on the Canvas that is passed in. You create the bit map, clear it, draw nothing into it, then draw the empty bitmap onto the main canvas.

**Comment by PhilJay** [comment]
I'm not sure tbh.
From my side I don't know what else to do.
Are you absolutely sure your own code is not leaking anywhere else?
Again: OOM-Exception just means that there was not enough memory to allocate the needed space. The cause of OOM is not neccessarily the cause of a (the) memory leak.
A perfectly fine component can cause an OOM even though somewhere else is the actual memory leak.

**Comment by brcinho** [comment]
Submitted a pull request, it is my first :smile: so if I did something wrong let me know.

As I wrote in the comment, I don't know which code change contributed more, whether settings canvas and its reference to bitmap to null or pinging system to perform garbage collection.
I don't know also if calling gc() is a bad thing and can do something wrong. As I understood, it is an instruction to jvm that gc() should be performed, and not an explicit call to garbage collection, so maybe it really can not do any harm. Your call.

Results can be observed here performed by manual testing by opening and closing Line Chart (or Pie Chart) form in example app around 50 times as fast as possible:

Before changes:

![before_changes](https://cloud.githubusercontent.com/assets/5712784/14510934/63f23034-01d4-11e6-9a5f-15222e156d44.png)
 

After changes:

![after_changes](https://cloud.githubusercontent.com/assets/5712784/14510944/6f188f76-01d4-11e6-9d8c-32f12ffac1ba.png)
